README
=head1 INTRODUCTION

Please see the INSTALLATION section below for build instructions.

This module provides access to the administrative functions of DB2 for
Linux, Unix and Windows.  The administrative functions are those
functions not accessible through SQL, which normally require use of
the DB2 command line processor (the "db2" command).  IBM has been
making some of these available through administrative view and the
SYSPROC.ADMIN_CMD stored procedure.

Using these functions from perl is faster and can provide more
detailed results than the command-line processor.  For example, when
reading configuration parameters, the administrative API will not just
tell you which values are set to "automatic", but also which options
support the "automatic" setting.  In a similar manner, when loading
data into a DPF database, the API returns detailed information for
each node and agent, allowing far more fine-grained error handling.

The DB2 administrative function module provides the following
functionality:

=over

=item *

Monitoring: get snapshot information and read event monitor files

=item *

Bulk data transfer: export, import, load

=item *

Configuration: get/set instance and database level configuration parameters

=item *

Directory: read or modify the database, node and DCS directory

=item *

Utilities: force applications, runstats, rebind, list history, list utilities

=back

=head1 PLATFORMS

This package has been known to build with the following combinations
of software.  Please report problems and additional platform / DB2
release combinations to the maintainer.

  Operating System        DB2 Version        Perl Version
  ================        ===========        ============
  AIX (unknown release)	  8.2		     5.8
  Red Hat EL 3 (x86)      8.1                5.8
  Red Hat EL 3 (x86)      8.2                5.8
  Red Hat EL 3 (x86)      9.1                5.8
  Red Hat EL 3 (x86)      9.5                5.8
  Red Hat EL 4 (x86_64)   8.2                5.8
  Red Hat EL 4 (x86_64)   9.1                5.8
  Red Hat EL 4 (x86_64)   9.5                5.8
  Solaris 7 (SPARC)       7.2                5.8
  Solaris 8 (SPARC)       8.2                5.8
  Solaris 10 (SPARC)      9.1                5.8
  Solaris 10 (SPARC)      9.5                5.8

=head1 INSTALLATION

This module can only be installed if you have the C compiler used to
build perl itself, plus a version of DB2 with the header files
installed (the developer's edition).  You generally cannot build this
module if you've downloaded a pre-compiled version of perl.

You need to run the test portions under a userid with DB2
system administration privileges.

Make sure the pre-requisite module Params::Validate has been
installed.

As a first step, set the environment variable DB2_VERSION to the DB2
release you have installed.  For example, if you're running DB2 Viper
(V9.1), do:

  export DB2_VERSION=V9.1

If you're not on Linux, Solaris or AIX (or if you've installed DB2 in
a different location than the Makefile.PL file expects), you also have
to specify the path where the DB2 files can be found, for example:

  export DB2_PREFIX=/opt/IBM/db2/V9.1

The common environment variables DB2DIR and DB2_HOME can be used as
well.

Please report the default prefix for platforms that the author doesn't
have access to (i.e. not Linux, Solaris), so the module can support
them in the future.

Then do:

  perl Makefile.PL
  make

The test step requires you to create a SAMPLE database with a set of
tables used for the export, import and load tests.  The supplied
"create_test_db" script will create the SAMPLE database using the
IBM-supplied "db2sampl" utility and create some additional tables.
Make sure you have sourced the DB2 environment before running the
script.

  create_test_db
  make test

You can run the tests against any other database you want, but please
make sure you update the CONFIG file before running the test script.

Once the test completes, you can install the module with:

  make install

=head1 RELEASE NOTES

The CHANGES.html file has a complete, historical list of all
user-visible (and some invisible) changes to this code.

=head1 AUTHORS

The code is currently maintained and supported by:

  Hildo Biersma <Hildo.Biersma@MorganStanley.com>

=cut

#
# $Id: README,v 160.1 2008/05/28 18:32:34 biersma Exp $
# 
# Copyright (c) 2007-2008, Morgan Stanley & Co. Incorporated
# See ..../COPYING for terms of distribution.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
# General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301  USA
#
# THE FOLLOWING DISCLAIMER APPLIES TO ALL SOFTWARE CODE AND OTHER
# MATERIALS CONTRIBUTED IN CONNECTION WITH THIS DB2 ADMINISTRATIVE API
# LIBRARY:
#
# THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
# BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS
# SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING
# THIS OR ANOTHER EQUIVALENT DISCLAIMER AS WELL AS ANY OTHER LICENSE
# TERMS THAT MAY APPLY.


CHANGES
=head1 Changes in the 3.0 release

Add support for the Backup command, including support for DPF backups
(with included logs) in DB2 V9.5.

=head1 Changes in the 2.9 release

=head2 Fix core dump in load from cursor

A second bug in the db2Load command, caused a load from cursor (source
type "Statement") to perform an invalid free, sometimes leading to a
core dump. Now fixed.  Have run valgrind on import/export/load to
verify no additional mistakes are present.

=head1 Changes in the 2.8 release

=head2 Fix core dump in load from cursor

A bug in the db2Load command, introduced in the 2.7 release, caused a
load from cursor (source type "Statement") to dump core. Now fixed.

=head2 BugFix in Load / Import

A bug in the Load and Import methods prevented load/import operations
against mixed-case schema and table names (which have to be
double-quoted when passed to DB2).  Now fixed.

=head2 Add the 'ReOpt' option to the Rebind method

The re-optimization level can now be set at Rebind time.

=head1 Changes in the 2.7 release

=head2 Rename module to DB2::Admin

The modules/classes have been renamed from DB2API to DB2::Admin as
recommended by Tim Bunce.

=head2 Add "list of files" support for Load of DEL files

The Load method now supports loading a list of DEL files in a single
command.

=head2 Support new protocol, types for node directory

Support the TCP/IP v4, TCP/IP v6 and TCP/IP v4 using SOCKS protocol
types for the node directory methods.

=head1 Changes in the 2.6 release

=head2 Add XML support to Export, Import and Load methods

The Export, Import and Load methods now supports the XML options
introduced in DB2 V9.1 (Export, Import) and DB2 V9.5 (Load).  Note
that the schema-related XML validate / default / ignore / map options
are not supported at this time.

=head2 Add connect attributes

Allow the user to specify the program name and connect time-out as
connect attributes.  Default options can be set with
C<SetConnectAttributes> and per-connect options can be specified with
the C<ConnectAttr> parameter to the C<Connect> method.

The default program name is no longer C<perl> or C<perl5.8>, but the
script name (the basename of C<$0>).

=head1 Changes in the 2.5 release

=head2 Add the NoCharDel option for the import and load commands

The NoCharDel option, which is used to handle files where character
strings have no delimiters, is now supported for FileOptions
parameters of the the Import and Load methods.  It was already
supported for the Import method.

=head2 Support "Update Profile Only" for Runstats

The implementation of the runstats API contained a bug that made
"update profile only" not work.  This was due to an discrepancy
between the documented constant name in the IBM documentation
(DB2RUNSTATS_UPDATE_PROFILE_ONLY) and the actual #define
(DB2RUNSTATS_UPDA_PROFILE_ONLY), missing the letters TE from the
middle.

The code now handles both cases

=head2 Fine-tune error handling

The error handling now differentiates between errors and warnings.
The default behavior is unchanged - error messages are printed for
errors and warnings, and no exception is thrown.  However, the
'RaiseError' option now only throws exceptions for errors, not
warnings - the new 'RaiseWarn' option handles the latter.

=head2 Support the ExcludingXML option for the Runstats command

The new DB2RUNSTATS_EXCLUDING_XML flag for the db2Runstats API is now
supported as the ExcludingXML option.

=head2 Support the CodePage file type modifier

The Export, Import, and Load commands now support the CodePage file
type modifier.  The Load command claimed to support this before,. but
the implementation was incorrect.

=head2 Rename the system admin group constants

The SysAdm, SysMaint, SysCtrl and SysMon group instance configuration
parameters (SQLF_KTN_xxx_GROUP) were limited to 16 characters in DB2
V8.1 and before.  With DB2 V8.2, new configuration parameters were
introduced that changed this limit to 30, but had a different name:
SQLF_KTN_xxx_GRP.  However, the documentation hides this difference
and refers to the new parameters by the old name.

DB2 V9.1 introduces a bug in the handling of the old parameters.  In
order to work around this, the new parameters now override the old
ones in the parse_constants script.  So in DB2 V8.2 and above, the old
names (e.g. 'sysadm_group') can be used and do the right thing by
using the new constants.

=head1 Changes in the 2.4 release

=head2 Changes in db2CfgGet / db2CfgSet for V9.1

The get/update instance/database configuration parameters methods have
new flags for DB2 V9.1: 'Computed' (get/set) and 'Manual' (set).  Add
support for these new flags.

The only value that supports 'Computed' as far as I am aware, is
'database_memory'.  When the Command Line Processor reports it as
'computed', the admin API can now do the same.

The 'Manual' option allow you to turn off automatic management for
parameters while DB2 maintains the last automatically set value.

=head2 Support for new format IP v4 for datastream elements

When IP v4 clients running DB2 V9.1 connect to a database, the
application id has a new format and includes the IP address and port
in plain text.  Support this new format and straighten out the
existing IP v4 format so it is now always "<address> port <port>".

=head1 Changes in the 2.3 release

=head2 Changes in db2Load API for V9.1

On DB2 V9.1, use the "iSetIntegrityPending" field for the db2Load API
instead of the "iCheckPending" field.

=head2 IP v6 support for datastream elements

Support IP v6 addresses for the application-id in snapshot and
event monitor data elements (untested)

=head2 Add new instance and database configuration parameters for V9.1

The following database manager and database configuration parameters
are new for DB2 V9.1 and have been added:

=over 4

=item federated_async

=item fcm_num_channels

=item self_tuning_mem

=item db_mem_thresh

=item db_collname

=item restrict_access

=back

=head1 Changes in the 2.2 release


util/parse_constants.pl
#!/usr/bin/perl
#
# parse_constants.pl - Parse DB2 <sqlmon.h> and <sqlutil.h> header
#                      files and create perl source code to to look
#                      up information and map numbers to strings
#
# $Id: parse_constants.pl,v 160.1 2008/05/29 19:12:08 biersma Exp $
#

use strict;
use Carp;
use Getopt::Long;

#
# Arguments:
# - basedir: DB2 base directory
# - outfile: name of output file
#
my %params = ('basedir' => '/opt/IBM/db2/V8.1',
	      'outfile' => 'lib/DB2/Admin/db2_constants.pl',
	     );
GetOptions(\%params, qw(basedir=s outfile=s)) ||
  die "Error parsing \@ARGV\n";
foreach my $req (qw(basedir outfile)) {
    next if (defined $params{$req});
    die "Missing required argument -$req\n";
}
die "-basedir $params{basedir}: no such directory\n"
  unless (-d $params{basedir});
my @db2_header_files = map "$params{basedir}/include/$_",
  qw(sqlmon.h sqlutil.h);
foreach my $fname (@db2_header_files) {
    die "$fname: no such file"
      unless (-f $fname);
}

my $constants = parse_header_files(@db2_header_files);
print "Generating code...";
generate_perl_code($constants, $params{'outfile'}, @db2_header_files);
print "ok\n";
exit(0);


#
# Parse the constants in the header files, and return a data structure
# mapping constant name to value and comment.  This ignores any
# ifdef/endif statements; if we need those, life will become more
# complex (aka the perl5 MQSeries module).
#
sub parse_header_files {
    my @header_files = @_;
    my %retval;

    foreach my $fname (@header_files) {
	open (HEADER, $fname) ||
	  die "Cannot open header file [$fname]: $!";
	my @lines = <HEADER>;
	chomp @lines;
	close(HEADER);

	while (@lines) {
	    my $line = shift @lines;
	    $line =~ s/^\s+//;
	    next unless ($line =~ /^#define/);

	    #
	    # Handle lines with continuations (a single backslash at
	    # the end of a line)
	    #
	    while (substr($line, -1) eq '\\') {
		$line =~ s!\s*\\$! !;
		my $next = shift @lines;
		$next =~ s/^\s+//;
		$line .= $next;
	    }
				
	    #
	    # We may have multi-line comments.  We want to integrate those
	    # into a single line for the parsing code.
	    #
	    if ($line =~ m~/\*(?!.*\*/)~) {
		#
		# Look forward until we find the closing line
		#
		my $count = 0;
		while (defined $lines[$count] &&
		       $lines[$count] =~ m!^\s+[^\*]+\s*$!) {
		    $count++;
		}
		if ($lines[$count] =~ m!^\s+[\(\w].*\*/\s*$!) {
		    #print "Have ", $count+1, "-line run-on comment, joining\n";
		    $lines[$_] =~ s!^\s+! ! foreach (0..$count);
		    $line .= shift @lines foreach (0..$count);
		} else {
		    die "Cannot parse multi-line comment starting at $line";
		}
	    }
				
	    my ($name, $value, $type, $comment);
	    if ($line =~ m!^\#define\s+\w+\s+/\*! ||
		$line =~ m!^\#define\s+\w+\s*$!) {
		#print "Skip empty constant: $line\n";
		next;
	    } elsif ($line =~ m!^#define\s+(\w+)\s+([^\/]+?)\s*$!) {
		#print "Constant with value but no comment: $line\n";
		($name, $value) = ($1, $2);
	    } elsif ($line =~ m!^#define\s+(\w+)\s+([^/]+?)\s*/\*\s*(.*?)\s*\*/\s*$!) {
		#print "Constant with value and C-style comment: $line\n";
		($name, $value, $comment) = ($1, $2, $3);
	    } elsif ($line =~ m!^#define\s+(\w+)\s+([^/]+?)\s*//\s*(.*?)\s*$!) {
		#print "Constant with value and C++-style comment: $line\n";
		($name, $value, $comment) = ($1, $2, $3);
	    } elsif ($line =~ m!^#define\s+(\w+)\(!) {
		#print "Skip macro with arguments: $line\n";
		next;
	    } else {
		die "ERROR: cannot parse line [$line]\n";
	    }
	    if (defined $retval{$name}) {
		die "Duplicate constant [$name]: maybe need #ifdef parsing";
	    }

	    #
	    # In V9.1, some comments start with a number,
	    # e.g. '07.List of connections to a DB'.  Remove such
	    # leading numbers.
	    #
	    if (defined $comment && 
		$comment =~ /^\d\d?\.([A-Z].*)$/s) {
		#print "Trim leading number of comment [$comment]\n";
		$comment = $1;
	    }

	    if ($value =~ m!^-?\d+$!) { # Decimal
		$type = 'Number';
	    } elsif ($value =~ m!^\((-?\d+)\)$!) { # Decimal, in parentheses
		$value = $1;
		$type = 'Number';
	    } elsif ($value =~ m!^0x[\da-fA-F]+$!) { # Hex
		$type = 'Number';
		$value = hex($value);
	    } elsif ($value =~ m!^'[^\']'$!) {
		$type = 'Character';
	    } elsif ($value =~ m!^"[^\"]+"$!) {
		$type = 'String';
	    } elsif ($value =~ m!^\(\w+\)\(?-?\d+\)?$!) { # Decimal with cast
		$type = 'Expression';
	    } elsif ($value =~ m!^sizeof\(\w+.*\)$!) { # Expression
		$type = 'Expression';
	    } elsif ($value =~ m!^\(\w+\s*[+-\|]\w+.*\)$!) { # Expression
		#print STDERR "Have complex expression [$value]\n";
		$type = 'Expression';
	    } elsif ($value =~ m!^\w+$!) {
		#
		# Hmmm - probably another constant
		#
		if (defined $retval{$value}) {
		    #print "Reference value: [$name] -> [$value]\n";
		    ($value, $type, $comment) =
		      @{ $retval{$value} }{qw(Value Type Comment)};
		} else {
		    #print "Skip reference to unknown constant [$value]: $line\n";
		    next;
		}
	    } else {
		die "Cannot parse constant [$value]: $line\n";
	    }
								
	    my $category;
	    if ($name =~ /^SQLM_ELM_/) {
		$category = 'Element';
	    } elsif ($name =~ /^SQLM_PLATFORM_/) {
		$category = 'Platform';
	    } elsif ($name =~ /^SQLM_TYPE_/ &&
		     $name ne 'SQLM_TYPE_HEADER') {
		$category = 'Type';
	    } elsif ($name =~ /^SQLM_CLASS_/) {
		$category = 'Class';
	    } elsif ($name =~ /^SQLM_HEAP_/ &&
		     $name ne 'SQLM_HEAP_MIN' && $name ne 'SQLM_HEAP_MAX') {
		$category = 'Heap';
	    } elsif ($name =~ /^SQLM_EVENT_/) {
		$category = 'Event';
	    } elsif ($name =~ /^SQLF_(?:DBTN|KTN)_/) {
		$category = 'ConfigParam';
	    } elsif ($name =~ /^SQL[FU]_RC_/) {
		$category = 'Reason';
	    } elsif ($name =~ /^SQLB_CONT_.*(?:PATH|DISK|FILE)$/) {
		$category = 'ContainerType';
	    } elsif (($name =~ /^SQLM_LO..$/ && $name ne 'SQLM_LOAD') ||
		     $name eq 'SQLM_LNON' ||
		     $name eq 'SQLM_LSIX') {
		$category = 'LockMode';
	    } elsif ($name =~ /^SQLM_\w+_LOCK$/) {
		$category = 'LockObjectType';
	    } elsif ($name =~ /^SQLM_(?:PREPARE|EXECUTE|EXECUTE_IMMEDIATE|OPEN|FETCH|CLOSE|DESCRIBE|STATIC_COMMIT|STATIC_ROLLBACK|FREE_LOCATOR|PREP_COMMIT|CALL|SELECT|PREP_OPEN|PREP_EXEC|COMPILE|SET|RUNSTATS|REORG|REBIND|REDIST|GETTA|GETAA)$/) {
		$category = 'StatementOperation';
	    } elsif ($name =~ /^SQLM_(?:STATIC|DYNAMIC|NON_STMT|STMT_TYPE_UNKNOWN)$/) {
		$category = 'StatementType';
	    } elsif ($name =~ /^SQLM_\w+_TABLE$/) {
		$category = 'TableType';
	    } elsif ($name =~ /^SQLM_UTILITY_\S+$/ &&
		     $name !~ /^SQLM_UTILITY_(STATE|INVOKER)_/ &&
		     $name ne 'SQLM_UTILITY_UNTHROTTLED') {
		$category = 'UtilityType';
	    } elsif ($name =~ /^SQLM_HADR_ROLE_/) {
		$category = 'HadrRole';
	    } elsif ($name =~ /^SQLM_HADR_STATE_/) {
		$category = 'HadrState';
	    } elsif ($name =~ /^SQLM_HADR_CONN_/) {
		$category = 'HadrConnectStatus';
	    } elsif ($name =~ /^SQLM_HADR_SYNCMODE_/) {
		$category = 'HadrSyncMode';
	    } elsif ($name =~ /^SQLM_REORG_(?:STARTED|PAUSED|STOPPED|COMPLETED|TRUNCATE)$/) {
		$category = 'ReorgStatus';
	    }

	    #
	    # For self-describing data stream elements,
	    # use the name as a back-up comment
	    #
	    if (! defined $comment &&
		$name =~ /^SQLM_ELM_(\w+)$/) {
		$comment = ucfirst(lc $1);
		$comment =~ tr/_/ /;
	    }
								
	    $retval{$name}{'Type'} = $type;
	    $retval{$name}{'Value'} = $value;
	    $retval{$name}{'Comment'} = $comment if (defined $comment);
	    $retval{$name}{'Category'} = $category if (defined $category);
	}			# End while: @lines
    }				# End foreach: header file
    print "Completed parsing\n";

    #
    # DB2 switched the various system administratore groups to new
    # configuration elements.  The old ones end on '_GROUP', the new
    # ones end on '_GRP', The documentation still says '_GROUP'.
    #
    # In DB2 V9, this makes a difference, as the code for the old
    # elements doens't handle values over 8 characters correctly.
    # So we replace the '_GROUP' values by the '_GRP' values.
    #
    foreach my $group (qw(SYSADM SYSMAINT SYSCTRL SYSMON)) {
	my $old = "SQLF_KTN_${group}_GROUP";
	my $new = "SQLF_KTN_${group}_GRP";
	if (defined $retval{$old} && defined $retval{$new}) {
	    $retval{$old} = delete $retval{$new};
	}
    }

    #
    # For upwards compatibility (and to make the main API work), we
    # define this V8 constant in lower releases
    #
    unless (defined $retval{'SQLM_CLASS_DEFAULT'}) {
	$retval{'SQLM_CLASS_DEFAULT'} =
	  { 'Type'    => 'Number',
	    'Value'   => 0,
	    'Comment' => 'SQLMA is for a standard snapshot',
	  };
    }

    return \%retval;
}


#
# Generate perl code with constant information in hashes.  This can be
# included (using 'require') by another module that provides an API.
#
sub generate_perl_code {
    my ($constants, $outfile, @header_files) = @_;

    my $file_names = join("\n", map "#    $_", @header_files);

    open (OUTPUT, "> $outfile") ||
      die "Cannot open output file [$outfile]: $!";

    print OUTPUT <<_END_;
#
# WARNING: This file is automatically generated.
# Any changes made here will be mercilessly lost.
#
# You have been warned, infidel.
#
# The file has been generated based on the following IBM header file:
#
$file_names
#
# and for the evil hackery used to generate this, see:
#
#    ..../util/parse_constants.pl
#

_END_
  ;

    #
    # Generate hash with all constant info, in alphabetical order
    #
    print OUTPUT "\$constant_info = {\n";
    foreach my $name (sort keys %$constants) {
	print OUTPUT ' ' x 4, "'$name' => {\n";
	my $data = $constants->{$name};
	foreach my $attr (sort keys %$data) {
	    my $val = $data->{$attr};
	    $val =~ s!'!\\'!g;
	    print OUTPUT ' ' x 8, "'$attr' => '$val',\n";
	}
	print OUTPUT ' ' x 4, "},\n";
    }
    print OUTPUT "};\n\n";

    #
    # Generate two-dimensional hash mapping category
    # to number to name
    #
    my %index;
    while (my ($name, $info) = each %$constants) {
	my ($type, $cat, $value) = @{$info}{qw(Type Category Value)};
	next unless (defined $cat && $type eq 'Number');
	$index{$cat}{$value} = $name;
    }

    print OUTPUT "\$constant_index = {\n";
    foreach my $category (sort keys %index) {
	my $l1 = $index{$category};
	print OUTPUT ' ' x 4, "'$category' => {\n";
	foreach my $value (sort { $a <=> $b } keys %$l1) {
	    print OUTPUT ' ' x 8, "'$value' => '$l1->{$value}',\n";
	}
	print OUTPUT ' ' x 4, "},\n";
    }
    print OUTPUT "};\n\n";

    print OUTPUT "1;\n";
}


Makefile.PL
#
# MakeMaker file for the DB2 Administrative API
#
# $Id: Makefile.PL,v 160.2 2008/05/28 18:31:44 biersma Exp $
#

use Config;
use ExtUtils::MakeMaker;
use strict;

#
# DB2_VERSION will be 7.1, 7.2, 8.1, 8.2, 9.1, 9.5
#
my $db2_version = $ENV{DB2_VERSION} || do {
    #
    # Exit with zero status to avoid warnings from CPAN testers
    #
    warn "Environment variable 'DB2_VERSION' not set\n";
    exit(0);
};

my ($db2_version_enum, @defines, %extra_args);

#
# Determine version enum
#
if ($db2_version =~ /^V7\./) {
    $db2_version_enum = 'db2Version710';
} elsif ($db2_version =~ /^V8\./) {
    $db2_version_enum = 'db2Version810';
} elsif ($db2_version =~ /^V9\.1/) {
    $db2_version_enum = 'db2Version900';
} elsif ($db2_version =~ /^V9\.5/) {
    $db2_version_enum = 'db2Version950';
} else {
    die "Unsupported DB2 version '$db2_version'\n";
}
push @defines, "-DDB2_VERSION_ENUM=$db2_version_enum";

#
# Determine path.  We look at the following environment variables:
# - DB2DIR
# - DB2_HOME
# After that, we fall back to platform defaults (guesses)
#
# All this can be overridden by DB2_PREFIX for Morgan Stanley hacks,
# see below.
#
my $db2_path;
foreach my $env_var (qw(DB2DIR DB2_HOME)) {
    next unless (defined $ENV{$env_var} && -d $ENV{$env_var});
    print "Using \$$env_var to set DB2 location\n";
    $db2_path = $ENV{$env_var};
    last;
}
unless (defined $db2_path) {
    if ($db2_version =~ /^V7\./) {
	$db2_path = "/opt/IBMdb2/V7.1";
    } elsif ($db2_version =~ /^V8\./) {
	if ($^O eq 'linux' || $^O eq 'solaris') {
	    $db2_path = "/opt/IBM/db2/V8.1";
	} elsif ($^O eq 'aix') {
	    $db2_path = "/usr/opt/db2_08_01";
	}  else {
	    die "unexpected V8.2 platform '$^O'";
	}
    } elsif ($db2_version =~ /^V9\.1/) {
	if ($^O eq 'linux') {
	    $db2_path = "/opt/ibm/db2/V9.1";
	} elsif ($^O eq 'solaris') {
	    $db2_path = "/opt/IBM/db2/V9.1";
	} elsif ($^O eq 'aix') {
	    $db2_path = "/usr/opt/db2_09_01";
	} elsif ($^O eq 'MSWin32' || $^O eq 'MSWin64') {
	    $db2_path = "C:\\ibm\\sqllib";
	} elsif ($^O eq 'cygwin') {
	    $db2_path = "/cygdrive/c/ibm/sqllib";
	} else {
	    die "unexpected V9.1 platform '$^O'";
	}
    } elsif ($db2_version =~ /^V9\.5/) {
	if ($^O eq 'linux') {
	    $db2_path = "/opt/ibm/db2/V9.5";
	} elsif ($^O eq 'solaris') {
	    $db2_path = "/opt/IBM/db2/V9.5";
	} elsif ($^O eq 'aix') {
	    $db2_path = "/usr/opt/db2_09_05";
	} elsif ($^O eq 'MSWin32' || $^O eq 'MSWin64') {
	    $db2_path = "C:\\ibm\\sqllib";
	} elsif ($^O eq 'cygwin') {
	    $db2_path = "/cygdrive/c/ibm/sqllib";
	} else {
	    die "unexpected V9.5 platform '$^O'";
	}
    } else {
	die "Unsupported DB2 version '$db2_version'\n";
    }
}

#
# This deals with a Morgan Stanley specific setup, where we install
# DB2 in a non-standard location and then make /opt/IBM/xxx link to
# it.  If $ENV{DB2_PREFIX} is set and resolves to a directory, use that
# as the db2path for compilation/installation.
#
# We also do some horrible link-time tricks.  Don't try this at home...
#
if ($ENV{DB2_PREFIX}) {
    $db2_path = $ENV{DB2_PREFIX};
    die "DB2_PREFIX path '$db2_path' invalid" unless (-d $db2_path);

    #
    # For DB2 V9.1, we have 'lib32' and 'lib64' but no 'lib'.
    #
    # FIXME: we may also need this for 32-bit, and we will
    # need to change this for 64-bit versions of perl.
    #
    # FIXME: We really should do this based on the DB2 release and the
    # bitness of perl.
    #
    my $lib_pf = 'lib';
    if ($ENV{ID_EXEC} eq 'ia32.linux.2.6.glibc.2.3' ||
	$ENV{ID_EXEC} eq 'x86_64.linux.2.6.glibc.2.3') {
	$extra_args{CCFLAGS} = '-m32';
	$lib_pf = 'lib32';
    } elsif ($ENV{ID_EXEC} eq 'sun4.sunos64.5.10') {
	$lib_pf = 'lib32';
    }

    my $sysliblist = "-L$db2_path/$lib_pf -ldb2";
    my @db2libs = qw(db2install db2g11n db2locale db2osse
		     db2genreg db2trcapi db2dascmn db2osse_db2);
    if ( $Config{osname} =~ /linux/i ) {
	$extra_args{LDDLFLAGS} = 
	  "-shared -Wl,--enable-new-dtags -Wl,-rpath $db2_path/$lib_pf";
	if ($ENV{ID_EXEC} eq 'ia32.linux.2.6.glibc.2.3' ||
	    $ENV{ID_EXEC} eq 'x86_64.linux.2.6.glibc.2.3') {
	    $extra_args{LDDLFLAGS} .= ' -m32';
	} else {
	    $sysliblist .= ' -L/tmp -lcxa';	# Setup in Makefile.inner
	}
    } elsif ( $Config{osname} =~ /solaris/i ) {
	#$sysliblist = "-R$db2_path/lib $sysliblist";
    }
    $sysliblist .= ' ' .join(' ', map {"-l$_"} @db2libs);
    $extra_args{LIBS} = [ $sysliblist ];
}

#
# In version V8.1 and above, we have the database / database
# manager configuration calls and db2Runstats.
#
if ($db2_version ge 'V8.1') {
    push @defines, "-DADMIN_API_HAVE_DB2BACKUP";
    push @defines, "-DADMIN_API_HAVE_DB2CFG";
    push @defines, "-DADMIN_API_HAVE_DB2RUNSTATS";
}

#
# In version 8.2 (but not above), use the V8.2 version db2DbDirXXX
# functions to access the database directory.
#
if ($db2_version eq 'V8.2') {
    push @defines, "-DADMIN_API_HAVE_DB2DBDIR_V8";
}

#
# In version 8.2 and above, use the db2Import/db2Export/db2Load
# functions for data import/export/load.
#
if ($db2_version ge 'V8.2') {
    push @defines, "-DADMIN_API_HAVE_DB2DBDIR";
    push @defines, "-DADMIN_API_HAVE_DB2EXPORT";
    push @defines, "-DADMIN_API_HAVE_DB2IMPORT";
    push @defines, "-DADMIN_API_HAVE_DB2LOAD";
}

#
# In version 9.1 and above, use the V9 version of the db2DbDirInfo
# structure and support XML for import/export
#
if ($db2_version ge 'V9.1') {
    push @defines, "-DADMIN_API_HAVE_DB2DBDIR_V9";
    push @defines, "-DADMIN_API_HAVE_EXPORT_XML";
    push @defines, "-DADMIN_API_HAVE_IMPORT_XML";
}

#
# In version 9.5 and above, use the LOngActionString for import
# and load; support XML for load.
#
if ($db2_version ge 'V9.5') {
    push @defines, "-DADMIN_API_HAVE_IMPORT_LONG_ACTION";
    push @defines, "-DADMIN_API_HAVE_LOAD_LONG_ACTION";
    push @defines, "-DADMIN_API_HAVE_LOAD_XML";
}

#
# Platform/bit-ness issues:
# - On Unix, link to libdb2.so; on Windows, link to db2api and db2api.
# - Use one of lib64/lib32/lib
#
my $libs = "-L$db2_path/lib -ldb2";  # Default
if ($^O eq 'MSWin32') {  # 32-bit Windows
    $libs = "-L\"$db2_path\\lib\" -ldb2cli -ldb2api";
} elsif ($^O eq 'MSWin64') {	# 64-bit Windows
    $libs = "-L\"$db2_path\\lib64\" -ldb2cli -ldb2api";
} elsif ($^O eq 'cygwin') {
    my $libdir = ($Config{use64bitall}) ? 'lib64' : 'lib32';
    $libs = "-L$db2_path/$libdir -ldb2cli -ldb2api";
} else {			# Unix
    my $libdir = ($Config{use64bitall}) ? 'lib64' : 'lib32';
    if (-d "$db2_path/$libdir") {
	$libs = "-L$db2_path/$libdir -ldb2";
    }
}
#print "XXX: have libs '$libs'\n";

WriteMakefile('NAME'         => 'DB2::Admin',
	      'VERSION_FROM' => 'lib/DB2/Admin.pm',
	      'PREREQ_PM'    => { 'Params::Validate' => '0.87',
				  'Test::Simple'     => '0.62',
				},
	      'INC'          => "-I$db2_path/include",
	      'LIBS'         => $libs,
	      'DEFINE'       => "@defines",
	      %extra_args,
	     );
